### ISSUE TITLE
About changes in the default language file and its translations

### ISSUE THREAD

**Issue by TechXavAL** [Beep6581/RawTherapee#6591]
@Beep6581 , @Entropy512 

It seems the final version for the `default` language file has been released (as of commit `e5007c77c`).

After reviewing the entries added at the end of the file `Espanol (Castellano)`, I have a couple of questions:

- it is advised we translate several missing texts, however in those related to PixelShift:
`!HISTORY_MSG_446;EvPixelShiftMotion`
`!HISTORY_MSG_447;EvPixelShiftMotionCorrection`
`!HISTORY_MSG_448;EvPixelShiftStddevFactorGreen`
`!HISTORY_MSG_450;EvPixelShiftNreadIso`
`!HISTORY_MSG_451;EvPixelShiftPrnu`
`!HISTORY_MSG_454;EvPixelShiftAutomatic`
`!HISTORY_MSG_455;EvPixelShiftNonGreenHorizontal`
`!HISTORY_MSG_456;EvPixelShiftNonGreenVertical`
`!HISTORY_MSG_458;EvPixelShiftStddevFactorRed`
`!HISTORY_MSG_459;EvPixelShiftStddevFactorBlue`
`!HISTORY_MSG_460;EvPixelShiftGreenAmaze`
`!HISTORY_MSG_461;EvPixelShiftNonGreenAmaze`
`!HISTORY_MSG_463;EvPixelShiftRedBlueWeight`
`!HISTORY_MSG_466;EvPixelShiftSum`
`!HISTORY_MSG_467;EvPixelShiftExp0`
`!HISTORY_MSG_470;EvPixelShiftMedian3`
it seems they are unused in the main (`default`) file. Wouldn’t it be more logical to just remove them (at least in the translated file)? If not, how should they be translated? That is, if they are not used, they won’t ever be shown in the GUI, and if in a future they are used for other purposes, such as for other tool, then the translation would become not only obsolete, but simply wrong…

- the last untraslated entry is a commented line:
`!//TP_LOCALLAB_SHOWMASKTYP_TOOLTIP;Mask and modifications can be chosen.\nBlur and noise : in this case it is not used for 'denoise'.\nDenoise : in this case it is not used for 'blur and noise'.\n\nBlur and noise + denoise : mask is shared, be careful to 'show modifications' and 'scope'`
which also won’t be shown in the GUI. So should it be translated or not?

Note this is for the translation to Castellano, but I don't know if there are other problems in other translations.

**Comment by TechXavAL** [comment]
Moreover, making a diff from version `5.8-3091-gfeada4cd0` (which is the base text used for the Castellano text), and quickly skimming over it, there are more discrepancies not pointed as having to be revised.

One particularly stricking is this one:

**was**: `EXPORT_USE_FAST_PIPELINE_TIP;`

**now is**: `EXPORT_USE_FAST_PIPELINE_TOOLTIP;`

and it's not pointed as a non translated entry. The result: it won't show the translation in the GUI, as the new entry is inexistent in Spanish.

**Comment by Lawrence37** [comment]
The commented entry was deleted from default and should be removed from the other language files. The changed language key names need to be changed in the other files as well. That's something I wanted to mention in the pull request but forgot to do. Finally, the unused history messages are there as a reminder that they are indeed unused rather than mistakenly left out or deleted. If it were up to me, I would translate the "--unused--"s too to make it clear that those entries have been dealt with.

**Comment by TechXavAL** [comment]
Fine.

Comented entry removed. Although there are many examples of keys just deleted without user interaction. This one should also have been deleted automatically.

     The changed language key names need to be changed in the other files as well

Don't know what you're referring to, sorry.

If it's about the change from **TIP** to **TOOLTIP** in quite a few keys, then it should have been done automatically, as those are keys and will never be translated (so safe to make a search and replace). If there's no warning about this in the translation file, most probably it won't be edited by translators. Anyway, already edited in the Castellano translation.

About the unused history messages: they were deleted indeed, but I have added them again, with `--unused--` translated. However, there are multitude of examples where entries have already been removed from `default` without notice (previous entries with proper translation, or obsolete entries, just deleted). And I feel obliged to ask for coherence: either all of the «not used» entries are deleted, or all of them are tagged as `--unused--`. And always warning the translators.

Finally, just one more thing: RT is UTF enabled, so may the Spanish languages be called «Español», «Español (Castellano)» y «Español (Latinoamericano)», please? (Note de letter **_ñ_** in _Español_)

**Comment by Entropy512** [comment]
As far as doing things automatically, improvements to the tooling is a possibility for the next release.

An alternative thing that is being discussed as a possible option is to investigate moving from a homegrown localization engine to GNU Gettext, as gettext has improved vastly since the last time someone made a decision along those lines according to @Beep6581 - in addition so has the tooling ecosystem, for example allowing translators to use something like weblate.

At the moment I'm personally leaning towards the second option, but my opinion of that might change if I do further investigation and find issues I'm unaware of as of yet.

I think some of the issue here may be that due to the extremely long time between releases and other people trying to pick up some of the default-cleanup tasks without ever having touched it before, some translation work wound up being done before the big generateTranslationDiffs run, and as a result, there were deltas in the translations from default that the tooling wasn't designed to handle.

**Comment by TechXavAL** [comment]
I was a Poedit happy user in the past, so following the gettext path would be wise, in my opinion.

However, no tool is perfect, and it won't do anything you haven't configured it to do: it won't take coherent decissions if it's not properly configured with the decissions taken by the developers. It also won't warn the translators about anything if not properly configured. And so on.

Let's just hope the rest of the translators realize all the work they should do to correctly update their work.

**Comment by Lawrence37** [comment]
@TechXavAL My comment about the key names was about the developers needing to rename them in all the language files. Once this is done, there is a script we can run to synchronize the translations with the updated default file. That should clear up any confusion, I hope.

**Comment by Beep6581** [comment]
> It seems the final version for the `default` language file has been released (as of commit `e5007c77c`).
> (...)
>       it seems they are unused in the main (`default`) file. Wouldn’t it be more logical to just remove them (at least in the translated file)?

Hey @TechXavAL , the `default` file is ready (or as ready as I wanted it to be for now - I skipped `LOCALLAB` as digging through that would be a tremendous job and it's not for 5.9), but the other language files have not been processed yet. I'll do that within a day.

**Comment by Beep6581** [comment]
Handled by #5664

**Comment by Lawrence37** [comment]
> Finally, just one more thing: RT is UTF enabled, so may the Spanish languages be called «Español», «Español (Castellano)» y «Español (Latinoamericano)», please? (Note de letter ñ in Español)

I'll find some time to try this out.

**Comment by Beep6581** [comment]
I suppose it could, but what if Japanese is renamed to 日本語, how will I type that in the console? :]

**Comment by TechXavAL** [comment]
Well, that's one of the challenges of internationalization. You can do it right, or you can fall by the wayside.

Let's see it as if I create a Spanish project and then add the English language, but call it _Inglis_ in the GUI...

Of course you can do whatever you feel better. It was more like a suggestion.

**Comment by Entropy512** [comment]
Hmm, perhaps keep it as-is for 5.9, and then look at further improvements afterwards?

Especially since there has been some discussion of nuking and repaving the entire system in favor of something else (possibly gettext) given the various maintainability issues it has.

It appears darktable uses country codes (not sure if this is standard for gettext?), haven't looked at other gettext examples.

**Comment by TechXavAL** [comment]
@Entropy512 

My gut feeling says it has nothing to do with the translated files, and something to do with a hardcoded pair of lines where the languages names shown in the GUI are set.

And I would bet that those lines are present in `/rtgui/multilangmgr.cc`, where you can read:

`emplace (key ("es", ""  ), "Espanol (Latin America)");`
`emplace (key ("es", "ES"), "Espanol (Castellano)");`

I wonder how hard would it be rewriting them to:

`emplace (key ("es", ""  ), "Español (Latinoamericano)");`
`emplace (key ("es", "ES"), "Español (Castellano)");`

This is a closed issue, but as long as nobody says this discussion is definitely finished, I will keep asking for the correctness of the Spanish texts. If I wasn't concerned with this, I wouldn't even had translated the whole GUI at all.

I can share quite a few examples where changing an «**ñ**» with an «**n**» changes the meaning of the word completely. And most Spanish people reading _Espanol_ in a program won't take it seriously right from the start.

**Comment by Lawrence37** [comment]
I have verified that RawTherapee can use unicode characters for languages. However, the name of the translation files determines what is shown in preferences and is also used in the options file to indicate which language has been set. If people are ok with resetting their preferred language, then we can implement the proposed improvement easily. Otherwise, there needs to be code to convert the old language names, or even better, don't use the file names for the text displayed in preferences.

**Comment by TechXavAL** [comment]
@Lawrence37 

Well, it seems you're right. By just changing the filenames and recompiling without updating from github, it works here without any more changes.

It is true I had to reset my preferred language in preferences, though, but I don't see it as a problem, just an annoyance to some people.

**Comment by Entropy512** [comment]
> It is true I had to reset my preferred language in preferences, though, but I don't see it as a problem, just an annoyance to some people.

Shouldn't it be automatically selected/detected from the system settings though?

see LocaleToLang in multilangmgr.cc

Of course naming the files by locale code (like gettext does) would eliminate the need to have that mapping.

The display in the UI could be handled by making the name of the language a translatable string in the translations engine - but it's going to be a pretty rare corner case that someone selects language in the UI as opposed to using system locale settings unless something else is broken.

**Comment by TechXavAL** [comment]
Well, I don't know exactly how it works. I thought it was there where the «magic» happened, however it seems it's not.

I will rephrase again just in case I haven't explained it well:
- I had only changed the filenames with a correct Spanish spelling and recompiled RT
- in _Preferences > Select language_ the correct Spanish spelling is shown, and you can select the proper language, which is correctly shown in the GUI with all UTF characters after restarting the program
- in the _options_ settings file, the correct spelling is set (the one present in the filename)
- I have no clue about what the _LocaleToLang_ section does in _multilangmgr.cc_ (if it does something, indeed)
- **I'm not asking** in any possible way that all this mess has to be fixed prior to releasing 5.9
- I would vote for changing to gettext sometime after releasing 5.9
- I'm just proposing an easy workaround so the spelling is correctly shown **in 5.9**

About the corner cases: I'm used to set my OS as English, with Spanish locale, and then set each app either to English or Spanish language, depending on how many (English) tutorials I have to watch to learn the app. Maybe a corner case, but I think I'm not alone doing this.

**Comment by Entropy512** [comment]
Aha, that might explain why you're not seeing the automatic language selection working correctly.

I suspect that in a "typical" use case, LocaleToLang SHOULD be automatically selecting the correct language file based on system settings.

In my case, on Linux:
declare -x LANG="en_US.UTF-8"

Which reminds me, I'm guessing that I might at some point need to do a translation run of my own, although US-isms are apparently so well known that Europeans are translating things and getting them right, since @heckflosse did the last round of changes to en_US.  :)

It does help finding out that I'm not actually in the default locale for translations, which might make it easier to test things when I attempt to migrate to gettext.  It LOOKS like it might be possible to have a msgid that is not the default translation text, which would make initial migration a lot easier.

**Comment by TechXavAL** [comment]
@Entropy512 

I'm sorry, but I think we are talking about different things: I'm not talking about any automatic selection, nor about recognizing system settings, nor about corner cases.

I'm just saying: if you go to _Preferences > Select language_ you will see an incorrect spelling of _Spanish_ in Spanish. And I'm asking if it's possible (indeed it is, quite easily) to change the filenames so that _Spanish_ (_Espanol_) is correctly writtten in Spanish (_Español_).

It is just a matter of changing 2 filenames, you have no need to change a single line of code, and has nothing to do with any other system setting of the operative system.

Of course, only if the dev team seems appropriate to do that.

**Comment by Lawrence37** [comment]
The code in `multilangmgr.cc` is for automatic language selection based on the system locale.

Resetting the language would only be necessary for those who have explicitly set it.

**Comment by TechXavAL** [comment]
Ok. So what I'm suggesting is changing these 2 lines of `multilangmgr.cc`:

from

```
emplace (key ("es", "" ), "Espanol (Latin America)");
emplace (key ("es", "ES"), "Espanol (Castellano)");
```

to

```
emplace (key ("es", "" ), "Español (Latinoamericano)");
emplace (key ("es", "ES"), "Español (Castellano)");
```

and changing the filenames:

from

```
Espanol (Latin America)
Espanol (Castellano)
```

to

```
Español (Latinoamericano)
Español (Castellano)
```

**Comment by Entropy512** [comment]
I think it would be better in the long term to decouple the filename from what is displayed in preferences for the override selection.  This allows solving @Beep6581 's concerns and yours at the same time.  Instead of the displayed string being the filename, the filenames would (ideally) just become the locale code (just like gettext, and in fact may be implemented by moving to gettext), and the string displayed in preferences would just be pulled from "yet another translation key" in the language file instead of from its name.

**Comment by TechXavAL** [comment]
`it would be better in the long term to decouple the filename from what is displayed in preferences`

Definitely!

Would it be done prior to releasing 5.9 version? If not, maybe the suggested workaround may work well just this time.

**Comment by Entropy512** [comment]
Unlikely, given that @Beep6581 was against even #6448 getting merged this late in the game, and that is a very low-risk item that fixes documented problems with camera support.  The cost/benefit of changing the translation file names is worse than that PR.

**Comment by TechXavAL** [comment]
@Entropy512 

I'm sorry, but right now I'm feeling either really dense, stupid, or I am clearly missing something important.

I'm understanding that changing 6 characters on the code and 6 characters on file names is even a risk?

If I'm correct, there was a point in time when `Espanol`(single entry) was replaced by `Espanol (Latin America)` and `Espanol (Castellano)` (two entries) in both the code and the filenames, and that didn't pose a risk. I would even say that there was absolutely not a single problem, but now there's a risk of breaking RT?

I say again that I've checked locally and there's no problem, RT works the same, no crashes. And even if there was a problem, the only thing broken would be the Spanish translation. RT would keep working the same.

But as I'm clearly missing something, could please elaborate in simple layman English?

Please, please?
